src/engine/authentication.md

```diff
@@ -0,0 +1,52 @@
+# Authentication
+
+The `engine` JSON-RPC interface, exposed by EL and consumed by CL, needs to be authenticated. The authentication scheme chosen for thus purpose is [JWT](https://jwt.io/). 
+
+The type of attacks that this authentication scheme attempts to protect against are the following: 
+
+- RPC port exposed towards the internet, allowing attackers to exchange messages with EL engine api. 
+- RPC port exposed towards the browser, allowing malicious webpages to submit messages to the EL engine api. 
+
+The authentication scheme is _not_ designed to 
+
+- Prevent attackers with capability to read ('sniff') network traffic from reading the traffic, 
+- Prevent attackers with capability to read ('sniff') network traffic from performing replay-attacks of earlier messages. 
+
+Authentication is performed as follows: 
+
+- For `HTTP` dialogue, each `jsonrpc` request is individually authenticated by supplying `JWT` token in the HTTP header. 
+- For a WebSocket dialogue, only the initial handshake is authenticated, after which the message dialogue proceeds without further use of JWT.
+  - Clarification: The websocket handshake starts with the client performing a websocket upgrade request. This is a regular http GET request, and the actual
+parameters for the WS-handshake are carried in the http headers. 
+- For `inproc`, a.k.a raw ipc communication, no authentication is required, under the assumption that a process able to access `ipc` channels for the process, which usually means local file access, is already sufficiently permissioned that further authentication requirements do not add security. 
+
+
+## JWT specifications
+
+- Client software MUST expose the authenticated Engine API at a port independent from legacy JSON-RPC API.
+  - The default port for the authenticated Engine API is `8551`. The Engine API is exposed under the `engine` namespace.
+- On the authenticated endpoint, the legacy API **MUST** also be available.
+- The EL **MUST** support at least the following `alg` `HMAC + SHA256` (`HS256`)
+- The EL **MUST** reject the `alg` `none`. 
+
+
+The HMAC algorithm means that symmetric encryption is used, thus several CL's will be able to use the same key, and, from an authentication perspective be able to impersonate each other. From a deployment perspective, it means that an EL does not need to be provisioned with individual keys for each caller. 
+
+## Key distribution
+
+The `EL` and `CL` clients **MUST** accept a cli/config parameter: `jwt-secret`, a `256` bit key, to be used for verifying/generating jwt tokens. 
+If such a parameter is not given, the client **SHOULD** generate such a token, valid for the duration of the execution, and show the token in the output, which the user can then use to provision the counterpart client with.
```

Comment:
Clients may alter this behaviour and persist the token when it was generated, right? It would help if machine is restarted and client launches upon the startup. Or we suppose this kind of setup requires pre-generated `jwt-secret`?

Reply:
Yes... So, problem: 
- If we output the secret in the logs, it would be good if the secret is temprorary. Users often supply log snippets in GH issues. It would not be good if a user unwittingly supplies a prod secret. 

Not sure if this is a good scheme. Maybe a better solution would be to do the same way geth does with the `nodeKey`: 

- If it doesn't exist, then create the file `nodeKey`, which is used as the enode secret in the future. 

Similarly, the node could spit out `jwtkey/hex`, containing the hex-encoded secret. So you boot up one of them, pick the `jwtkey.hex` file and feed it into the other client. That way nothing secret goes out the logs.